test/test_main.py

```diff
@@ -855,3 +863,46 @@ def _call_event(self, rel_path, env):
         dispatch = self.JUJU_CHARM_DIR / 'dispatch'
         subprocess.check_call([str(dispatch)],
                               env=env, cwd=str(self.JUJU_CHARM_DIR))
+
+
+class TestStorageHeuristics(unittest.TestCase):
+    def setUp(self):
+        logassert.setup(self, '')
+
+    def test_fallback_to_current_juju_version__too_old(self):
+        meta = CharmMeta.from_yaml("series: [kubernetes]")
+        with patch.dict(os.environ, {"JUJU_VERSION": "1.0"}):
+            self.assertFalse(_should_use_controller_storage(Path("/xyzzy"), meta))
+            self.assertLogged('Using local storage: JUJU_VERSION=1.0.0')
+
+    def test_fallback_to_current_juju_version__new_enough(self):
+        meta = CharmMeta.from_yaml("series: [kubernetes]")
+        with patch.dict(os.environ, {"JUJU_VERSION": "2.8"}):
+            self.assertTrue(_should_use_controller_storage(Path("/xyzzy"), meta))
+            self.assertLogged('Using controller storage: JUJU_VERSION=2.8.0')
+
+    def test_use_min_juju_version_if_set__too_old(self):
+        meta = CharmMeta.from_yaml("{series: [kubernetes], min-juju-version: '1.0'}")
```

Comment:
there is also the potential that there is no min juju version set.
which means 2.8 will create a stateful set. Should we use it?

Reply:
ah, i thought that was particular to one instance of it, not all 2.8s.
Is there a way to as juju if it's created a stateful set?

Reply:
You can't downgrade a deployment, so why don't we just check against the current version of juju rather than the metadata.yaml? ie. if we are deployed with juju>2.8, use controller storage.

Reply:
I think I'll go for this. Lines up better with the docstring, too.